Fix bug in graph partitioner and update graph signature after partitioning.

tugsbayasgalan opened the pull request:
Summary: This fix does two things:

1. When we add inputs from partioner to the top level graph module, we insert in the order of partioner which is not guaranteed to be same as original graph inputs. This PR fixes that. 
2. When we replace autograd ops with HOP, we create new submodules and access their outputs via getitem calls. As a result, previous node names associated with getitem gets updated, resulting in the graph being different from produced graph signature. So I just update the graph signature accordingly. 

Test Plan: CI

Differential Revision: D56688411

SherlockNoMad commented on torch/export/_trace.py:
```diff
@@ -534,13 +534,54 @@ def _compiling_state_context():
+        if node.op == "output":
+            updated_output_names = node.args
+
+    # It could be updated because replace_set_grad_with_hop_pass adds new
```
addd a test case for this?

tugsbayasgalan replied:
The internal diff has a torchrec unittest.

SherlockNoMad commented on torch/fx/passes/split_module.py:
```diff
@@ -477,6 +477,16 @@ def instantiate_node_partition_mapping(node):
                 orig_mod_env[key] for key in partition.inputs
             ]
+
+            # We actually need to insert the placeholder nodes in the original order
```
test case?

tugsbayasgalan replied:
The internal diff has a torchrec unittest.

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -534,13 +534,54 @@ def _compiling_state_context():
+            updated_output_names = node.args
+
+    # It could be updated because replace_set_grad_with_hop_pass adds new
+    # getitem nodes into the graph.
```
Right now it is only replace_set_grad_with_hop_pass, but in theory, it could be useful for any pass that adds submodules in the graph. I can make it a free function.

zhxchen17 commented on torch/_export/passes/replace_set_grad_with_hop_pass.py:
```diff
@@ -160,7 +160,58 @@ def _maybe_inline_or_replace_with_hop(node: torch.fx.Node):
+        if node.op == "output":
+            updated_output_nodes = node.args
+
+    # It could be updated because replace_set_grad_with_hop_pass adds new
+    # getitem nodes into the graph.
+    assert len(original_output_nodes) == len(updated_output_nodes)
+    old_output_name_to_new_output_name = {}
+    for k, v in zip(*original_output_nodes, *updated_output_nodes):
+        if isinstance(k, torch.fx.Node) and isinstance(k, torch.fx.Node):
+            old_output_name_to_new_output_name[k.name] = v.name
+        # If there is constant return in the end, it should be true for updated outputs too
+        else:
+            constant_types = (int, float, str, type(None))
+            assert k == v
+            assert isinstance(k, constant_types)
+
+    buffers_to_mutate_copy = graph_signature.buffers_to_mutate.copy()
+    user_inputs_to_mutate_copy = graph_signature.user_inputs_to_mutate.copy()
+    for k in old_output_name_to_new_output_name:
+        if k in graph_signature.buffers_to_mutate:
+            graph_signature.buffers_to_mutate[
+                old_output_name_to_new_output_name[k]
+            ] = buffers_to_mutate_copy[k]
+            if k not in old_output_name_to_new_output_name.values():
+                del graph_signature.buffers_to_mutate[k]
+
+        if k in graph_signature.user_inputs_to_mutate:
+            graph_signature.user_inputs_to_mutate[
+                old_output_name_to_new_output_name[k]
+            ] = user_inputs_to_mutate_copy[k]
+            if k not in old_output_name_to_new_output_name.values():
+                del graph_signature.user_inputs_to_mutate[k]
+
+    for i, k in enumerate(graph_signature.user_outputs):
+        if k in old_output_name_to_new_output_name:
+            new_k = old_output_name_to_new_output_name[k]
+            graph_signature.user_outputs[i] = new_k
```
I think you can get away these regeneration code with:
```
with graph_module._set_replace_hook(ep.graph_signature.get_replace_hook():
    _replace_set_grad_with_hop_pass(gm)
```

tugsbayasgalan replied:
Hmm we actually cannot use ep here because this is before we create ep. We need it before ep because ep verifier won't pass as we have non-functional ops in the graph at this point.

ydwu4 replied:
qq: is ep.graph_signature the same as the type as the graph signature here? If so, we could still use the method?

zhxchen17 replied:
discussed offline with @tugsbayasgalan, we should still use something more structural here. will follow up

zhxchen17 reviewed (changes requested):
sorry wrong button clicked lol

ydwu4 commented on torch/_export/passes/replace_set_grad_with_hop_pass.py:
```diff
@@ -160,7 +160,58 @@ def _maybe_inline_or_replace_with_hop(node: torch.fx.Node):
 
 
-def replace_set_grad_with_hop_pass(gm: torch.fx.GraphModule):
+def replace_set_grad_with_hop_pass(gm: torch.fx.GraphModule, graph_signature):
```
maybe a type annotation?

zhxchen17 replied:
yeah adding a type annotation here is better.

ydwu4 reviewed (approved):
Overall looks good to me. Left some minor comments on code structure.

zhxchen17 commented on torch/_export/passes/replace_set_grad_with_hop_pass.py:
```diff
@@ -141,35 +142,39 @@ def _sequential_split_and_maybe_inline_subgraphs(gm: torch.fx.GraphModule):
+            replace_ctx = new_gm._set_replace_hook(graph_signature.get_replace_hook())  # type: ignore[assignment]
+
+        with replace_ctx:
+            def _maybe_inline_or_replace_with_hop(node: torch.fx.Node):
+                if _is_set_grad_enabled_sub_mod(node, omit_if_same_with_ambient=True):
+                    _replace_with_hop(node)
+                else:
+                    _remove_set_grad_and_inline(node)
```
we can define this function out of the context manager.

zhxchen17 reviewed (approved):
some nits.

tugsbayasgalan committed:
Fix bug in graph partitioner (#125133)

Summary:
X-link: https://github.com/pytorch/torchrec/pull/1934


Title

Test Plan: CI

Reviewed By: PaulZhang12

Differential Revision: D56688411
